### PR TITLE
fix(dynamic-forms): prefix group keys on derivation paths

### PR DESCRIPTION
**Pull request by dereekb**
# Pull Request

## Description

Derivations attached to a value field nested inside a group (or under any number of groups, with optional layout containers in between) silently failed: the entry was registered with a bare fieldKey while the field's actual form path was group-prefixed, so the applicator's writeback target pointed at the form root and never landed. 

This PR fixes that. It also adds a $self and $group token so factory-built field shapes can self-target and group-target without knowing their ancestor group keys.

## Type of Change

<!-- Mark the relevant option with an "x" -->

- [x] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] 📝 Documentation update
- [ ] 🎨 Code style update (formatting, renaming)
- [ ] ♻️ Code refactoring (no functional changes)
- [ ] ⚡ Performance improvement
- [ ] ✅ Test update
- [ ] 🔧 Build/tooling update

## Related Issues

<!-- Link to related issues. Use "Fixes #123" to auto-close issues when PR is merged -->

Fixes #
Related to #

## Changes Made

- SELF_DEPENDENCY_TOKEN = '$self' exported from derivation/index.ts.
- resolveSelfDependencies() substitutes '$self' in dependsOn arrays with the field's resolved absolute path at collection time. Token is dependsOn-only — extractStringDependencies is unchanged, and a literal "$self" inside an expression string is not treated as a dependency.
- Works for shorthand and full-logic derivations, alongside other absolute deps, in nested groups, and inside arrays (resolves to the array placeholder form 'items.$.name').
- Added GROUP_DEPENDENCY_TOKEN = '$group' next to $self, exported from derivation/index.ts                                                                                                                                                                   
- buildEffectiveFieldKey now combines arrayPath + groupPath (fixes latent array > group > input bug — the form schema nests under the group, but the entry was using 'items.$.field' ignoring the group)                                                   
- New buildNearestGroupPath helper used by token resolution                                                                                                                                                                                                  
- resolveTokenDependencies (renamed from resolveSelfDependencies) substitutes both tokens AND their dotted prefix forms ($group.sibling, $self.X)
- Throws DynamicFormError when $group is used at form root
- Added groupValue?: unknown to EvaluationContext                                                                                                                                                                                                            
- createEvaluationContext populates it from the entry's parent path against formValue                                                                                                                                                                        
- createArrayItemEvaluationContext populates it from the array item's inner group path, falling back to the array item itself when the field is a direct child of the array          

- Fixes a latent bug in the collector where a derivation on a field nested
as `array > group > input` produced an entry path of `'items.$.field'`
(group ignored) instead of `'items.$.group.field'`. The form schema
correctly nests under the group, so writeback was silently failing the
same way the original group bug did, just inside an array. The collector
now combines arrayPath and inner groupPath into the entry path.

<!--
## Breaking Changes (if applicable)

**Before:**
```typescript
// Old usage
```

**After:**
```typescript
// New usage
```

**Migration guide:**
-
-->

## Testing

- [x] Unit tests added/updated
- [x] All tests pass locally
- [ ] Tested manually
- [ ] Tested edge cases
                                                                                                                                                                                                                                             
- Updated the existing collector test that asserted the buggy behavior (fullAddress → address.fullAddress).                                                                                                                                                  
- Added 4 collector tests for groupPath: prefixed groups, nested groups, group-with-container pass-through, and array boundary reset.
- Added 7 collector tests for $self: root, inside group, mixed with absolute paths, deeply nested, inside array, with HTTP derivation, with async function derivation, plus a defensive expression-extraction test.                                          
- Added two e2e scenarios container-inside-group and container-inside-group-self (apps/e2e/core/src/app/testing/container-nesting/) covering the group > container > input shape with absolute-path and $self variants.
                     
## Screenshots / Videos

<!-- If applicable, add screenshots or videos to demonstrate the changes -->

## Documentation

- [ ] Documentation updated (if needed)
- [ ] API documentation (TSDoc) added/updated
- [ ] Examples updated (if needed)

## Checklist

<!-- Ensure all items are checked before requesting review -->

- [x] Code follows the [coding standards](../CODING_STANDARDS.md)
- [x] No linting errors (`pnpm lint`)
- [x] Code is formatted (`pnpm lint:fix`)
- [x] Build succeeds (`pnpm build:libs`)
- [ ] Commits follow [conventional commit format](https://www.conventionalcommits.org/)
- [ ] Self-review completed
- [ ] No commented-out code or console.log() statements
- [ ] Type safety maintained (no `any` types)

## Additional Notes

<!-- Any additional information that reviewers should know -->

### CODE REVIEW

**Comment by dereekb** [comment]
The second commit summary:
                                                                                                                                                                                                                                                           
  Library (derivation-collector.ts):
  - Added GROUP_DEPENDENCY_TOKEN = '$group' next to $self, exported from derivation/index.ts                                                                                                                                                                   
  - buildEffectiveFieldKey now combines arrayPath + groupPath (fixes latent array > group > input bug — the form schema nests under the group, but the entry was using 'items.$.field' ignoring the group)                                                   
  - New buildNearestGroupPath helper used by token resolution                                                                                                                                                                                                  
  - resolveTokenDependencies (renamed from resolveSelfDependencies) substitutes both tokens AND their dotted prefix forms ($group.sibling, $self.X)
  - Throws DynamicFormError when $group is used at form root                                                                                                                                                                                                   
                                                                                                                                                                                                                                                               
  Runtime (derivation-applicator.ts, evaluation-context.ts):                                                                                                                                                                                                   
  - Added groupValue?: unknown to EvaluationContext                                                                                                                                                                                                            
  - createEvaluationContext populates it from the entry's parent path against formValue                                                                                                                                                                        
  - createArrayItemEvaluationContext populates it from the array item's inner group path, falling back to the array item itself when the field is a direct child of the array                                                                                  
                                                                                                                                                                                                                                                               
  Tests:                                                                                                                                                                                                                                                       
  - 6 new collector tests for $group (root, nested, array, array+group, mixed, throw)                                                                                                                                                                          
  - 5 new collector tests for prefix substitution ($group.X and $self.X, including nested groups, arrays, multi-segment, throw)                                                                                                                                
  - 1 new collector test for the array+group combine fix                                                                                                                                                                                                       
  - 3 new applicator tests for groupValue in non-array context (nested group, doubly-nested groups, root undefined)                                                                                                                                            
  - 1 new e2e scenario container-inside-group-parent using dependsOn: ['$group'] and ctx.groupValue end-to-end

File: packages/dynamic-forms/src/lib/core/derivation/derivation-collector.ts [hunk 1]
```diff
@@ -81,13 +100,16 @@ function traverseFields(fields: FieldDef<unknown>[], entries: DerivationEntry[],
...
-      const childContext = { ...context };
+      const childContext: CollectionContext = { ...context };
 
-      // Update array path context if this is an array field
       if (field.type === 'array') {
+        // Array boundary: the array placeholder path takes over and any
+        // ancestor groupPath is reset for descendants (entries inside
+        // array items use `arrayPath.$.fieldKey` form).
         childContext.arrayPath = field.key;
+        childContext.groupPath = undefined;
```
[the remark this answers is not in the document]
Comment (dereekb) [reply packages/dynamic-forms/src/lib/core/derivation/derivation-collector.ts]: Verified — this finding does not match the current design and should not be fixed.
                                                                                                                                                                                                                                                               
  Evidence the current behavior is intentional:                                                                                                                                                                                                                
                                                                                                                                                                                                                                                               
  1. JSDoc on buildEffectiveFieldKey (derivation-collector.ts:175-178) explicitly states:                                                                                                                                                                      
  ▎ groupPath is reset when entering an array (each array item is its own model root); ancestor groups OUTSIDE the array are not part of the entry path. Groups INSIDE the array are tracked via groupPath and combined here.                                  
  2. In‑code comment at derivation-collector.ts:129-131:                                                                                                                                                                                                       
  ▎ Array boundary: the array placeholder path takes over and any ancestor groupPath is reset for descendants (entries inside array items use arrayPath.$.fieldKey form).                                                                                      
  3. An explicit passing test at derivation-collector.spec.ts:255-282 titled "should reset ancestor groupPath at array boundaries (array preserves array.$.fieldKey form)" sets up exactly this shape — group:org > array:items > input:name — and asserts     
  collection.entries[0].fieldKey === 'items.$.name', not 'org.items.$.name'. Adopting the reviewer's proposed join(groupPath, field.key) would directly break this test.                                                                                       
  4. The reviewer's proposed example "address.items.$.name" is the opposite of what the spec encodes and contradicts the "each array item is its own model root" design principle.                                                                             
                                                                                                                                                                                                                                                               
  On the reviewer's framing: they claim composition "ignores ancestor groups". The composition isn't ignoring anything — buildEffectiveFieldKey deliberately does not concatenate arrayPath with an outer groupPath because the collector clears groupPath at  
  the array boundary by design. Lines 183-184 only combine arrayPath with groupPath when groups are inside the array (e.g., array > group > input → 'items.$.address.state'), which is covered by another passing test at derivation-collector.spec.ts:284-311.
                                                                                                                                                                                                                                                               
  No code change made. The finding is a misread of the intended semantics.

File: packages/dynamic-forms/src/lib/core/derivation/derivation-collector.ts [hunk 2]
```diff
@@ -106,12 +149,102 @@ function traverseFields(fields: FieldDef<unknown>[], entries: DerivationEntry[],
...
+ * - `GROUP_DEPENDENCY_TOKEN` ('$group') → the field's nearest parent
+ *   container path; throws if the field has no parent group/array.
+ * - `'$group.sibling'` → the parent path joined with `.sibling`, useful to
+ *   target a specific sibling without naming the parent group key.
+ *
+ * Non-token strings are passed through unchanged.
+ *
+ * @internal
+ */
+function resolveTokenDependencies(deps: string[], effectiveFieldKey: string, context: CollectionContext): string[] {
```
Comment (antimprisacaru) [inline packages/dynamic-forms/src/lib/core/derivation/derivation-collector.ts:187]: The HTTP and async-function guards reject empty `dependsOn` and `'*'` because either would cause request storms on every form change. `dependsOn: ['$group']` resolves at collection time to the parent path (e.g. `'address'`), which has the same blast radius as a group-scoped wildcard — the derivation fires whenever any sibling under that group changes. That's not necessarily wrong, but it sidesteps the rationale behind the existing guards.

Two options: (a) explicitly allow `$group`/`$group.X` for HTTP/async and document that the user owns the request frequency, or (b) reject `$group` (token form, not the resolved path) for HTTP/async sources alongside the wildcard check. I'd lean toward (a) for now with a TSDoc note here, but worth a deliberate call rather than letting it ship by accident.

**Review by antimprisacaru — COMMENTED** [review]
Looks good overall, though I see we intersected each other over a bit of refactoring I made earlier, sorry about the conflicts! 

I'll have another look once the conflicts are solved. Thanks for the great work!

File: packages/dynamic-forms/src/lib/core/derivation/derivation-collector.ts (hunk 2 above)
Comment (dereekb) [reply packages/dynamic-forms/src/lib/core/derivation/derivation-collector.ts:187]: Merge conflicts fixed. 

I updated the TSDocs to reflect that concern, but I think you're right, going option A falls more in-line with what is currently allowed. I'd rather keep the freedom there, rather than throwing an error or forcing the use of "trigger" and/or a near-0 debounce if the user really does want the form to fire each time the group changes somewhat.

**Comment by dereekb** [comment]
Found a few more bugs in testing. I think we're good now, but I'll keep testing it to find more edge cases.

I did see isRowField is still being used in a few places, but Claude noted that they're all used before the mapper so they're still useful/valid being used as-is (isRowField() won't always return false) but I'd suspect that it can also probably come out now that isRowField() is more of a Container/Wrapper now and probably shouldn't have a special case in these other types if Container doesn't have to be handled as a special case.

**Comment by dereekb** [comment]
Found one more, where:

-  fieldValue on a root-level item within an Array was returning the Array's entire value
- derivation was not triggering properly for the same case

Added E2E tests and derivation-applicator tests to cover this case

**Review by antimprisacaru — COMMENTED** [review]
I have to say, I really appreciate you digging deep into these things. You've really done a great deal in terms of stabilization and overall lib quality!

Moving on to the review, I wanted to say. While verifying that nothing relied on the old fieldValue === arrayItem behavior (it doesn't), I noticed property-derivation-applicator.ts:247–265 has the same createArrayItemEvaluationContext shape with the same fieldValue: arrayItem and fieldPath: items.0 issues. Worth fixing it in this PR for symmetry — $self and fieldValue are equally broken there, just on the property-derivation side. The two createArrayItemEvaluationContext functions are near-duplicates and could probably share an implementation.

**Comment by dereekb** [comment]
Yea, I'll have property-derivation files pull from the derivation files where possible. There are a few places identified where they can be simplified or use a shared utility.

**Comment by dereekb** [comment]
Oops, I might have merged in your other change accidentally. I'll fix that

**Comment by dereekb** [comment]
Ok, should be ready for review. Updated the property-derivation files to pull in/extend their related derivation types where applicable. Shaved a couple hundred lines of duplicate logic/interfaces.

**Review by antimprisacaru — APPROVED** [review]
